arb/sqr.c

```diff
@@ -0,0 +1,67 @@
+/*
+    Copyright (C) 2021 Albin Ahlbäck
+
+    This file is part of Arb.
+
+    Arb is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <http://www.gnu.org/licenses/>.
+*/
+
+#include "arb.h"
+
+void
+arb_sqr(arb_t res, const arb_t x, slong prec)
+{
+    mag_t resr, xm;
+    int inexact;
+
+    if (arb_is_exact(x))
+    {
+        inexact = arf_sqr_rnd_down(arb_midref(res), arb_midref(x), prec);
+
+        if (inexact)
+            arf_mag_set_ulp(arb_radref(res), arb_midref(res), prec);
+        else
+            mag_zero(arb_radref(res));
+    }
+    else if (ARB_IS_LAGOM(x) && ARB_IS_LAGOM(res))
+    {
+        mag_fast_init_set_arf(xm, arb_midref(x));
+
+        mag_init(resr);
+        mag_fast_mul(resr, xm, arb_radref(x));
+        if (resr->exp < COEFF_MAX && resr->exp >= COEFF_MIN)
```

Comment:
These checks are not needed; the exponent can be increment with risk of overflow.

Reply:
I presume you meant _without_  risk of overflow ;) Och god jul, Fredrik!

Reply:
God jul :)